### ISSUE TITLE
Executor/Code generators and line breaks within an instruction

### ISSUE THREAD

**Issue by GitMensch** [fesch/Structorizer.Desktop#416]
I've considered to add some line breaks for code imports but the first tries showed that this would break the execution. Simple example:

~~~
hex_tab <- "000102030405060708090a0b0c0d0e0f"  + "101112131415161718191a1b1c1d1e1f"  + "202122232425262728292a2b2c2d2e2f"  + "303132333435363738393a3b3c3d3e3f"  + "404142434445464748494a4b4c4d4e4f"  + "505152535455565758595a5b5c5d5e5f"  + "606162636465666768696a6b6c6d6e6f"  + "707172737475767778797a7b7c7d7e7f"  + "808182838485868788898a8b8c8d8e8f"  + "909192939495969798999a9b9c9d9e9f"  + "a0a1a2a3a4a5a6a7a8a9aaabacadaeaf"  + "b0b1b2b3b4b5b6b7b8b9babbbcbdbebf"  + "c0c1c2c3c4c5c6c7c8c9cacbcccdcecf"  + "d0d1d2d3d4d5d6d7d8d9dadbdcdddedf"  + "e0e1e2e3e4e5e6e7e8e9eaebecedeeef"  + "f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff" 
~~~

works in the execution but stretches the NSD window much too wide.

~~~
hex_tab <- "000102030405060708090a0b0c0d0e0f"
+ "101112131415161718191a1b1c1d1e1f"
~~~

looks better but leads to

> <+ "101112131415161718191a1b1c1d1e1f"> is not a correct function!

raised from `Executor.trySubroutine`

The reason is simple: the `Executor.stepInstruction` uses a `StringList` and splits the instructions (one per line) which we obviously can't change because of backwards compatibility.

Can I suggest to add a look-forward here? This would mean check the single elements and re-concatenate them with some "simple" rule-set (for example: add to the previous element when the element starts with an operator or the previous ended with an operator).

Note: just stumbled over https://github.com/codemanyak/Structorizer.Desktop/pull/79 when investigating this

**Comment by codemanyak** [comment]
I had come to the same point on import tests with some of the COBOL samples and pondered similar ideas to address the inconvenience. But I haven't found an easy solution so far. We must consider several side effects and implications (e. g. the transmutation between multiple single-line instruction elements and a multi-line instruction element).
(I just helped myself by manually splitting such superlong expressions into incremental assignments:
```
hex_tab <- "000102030405060708090a0b0c0d0e0f"
hex_tab <- hex_tab + "101112131415161718191a1b1c1d1e1f"
hex_tab <- hex_tab + "202122232425262728292a2b2c2d2e2f"
```
etc., which is a workaround for users but no solution of course.)
On the other hand, I'm not sure whether it's wise to add a solution for this problem to the agenda of relaese 3.27: A makeshift solution approach may cause more trouble than it is supposed to mend.
The following is not a real argument but should be weighed: How often will such enormously long lines occur? At least it had never been raised as an issue before we started to import COBOL sources (directly or indirectly via C).

**Comment by GitMensch** [comment]
The maybe easiest solution (while not that nice looking) would be to add another workaround we could use for the code imports and that people could use if they like to: use a line concatenator like C as this simplifies the rules down to one:
If the current line ends with "\" (or whatever symbol you like to use): concatenate it with the next line. This would lead to a still readable:

~~~
hex_tab <- "000102030405060708090a0b0c0d0e0f" \
         + "101112131415161718191a1b1c1d1e1f" \
         + "202122232425262728292a2b2c2d2e2f" \
         + "303132333435363738393a3b3c3d3e3f" \
         + "404142434445464748494a4b4c4d4e4f" \
         + "505152535455565758595a5b5c5d5e5f" \
         + "606162636465666768696a6b6c6d6e6f" \
         + "707172737475767778797a7b7c7d7e7f" \
         + "808182838485868788898a8b8c8d8e8f" \
         + "909192939495969798999a9b9c9d9e9f" \
         + "a0a1a2a3a4a5a6a7a8a9aaabacadaeaf" \
         + "b0b1b2b3b4b5b6b7b8b9babbbcbdbebf" \
         + "c0c1c2c3c4c5c6c7c8c9cacbcccdcecf" \
         + "d0d1d2d3d4d5d6d7d8d9dadbdcdddedf" \
         + "e0e1e2e3e4e5e6e7e8e9eaebecedeeef" \
         + "f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff" 
~~~

**Comment by codemanyak** [comment]
The draw method might force some minimum indentation for consecutive lines if the continuation lines don't already start indented.

**Comment by GitMensch** [comment]
Yes, this sounds reasonable. Thank you for the work on this.

**Comment by codemanyak** [comment]
For the line breaks it would be a fine heuristics to combine the existence with line breaks in the origibal code with some configurable line limits in Structorizer. Unformtunately the reduction tree doesn't contain the original line breaks anymore as these are filtered as "noise" (whitespace).
Nevertheless I will reconsider this.

**Comment by GitMensch** [comment]
>  Unfortunately the reduction tree doesn't contain the original line breaks anymore as these happen to be filtered as "noise" (whitespace).
> Nevertheless I will reconsider this.

A simple heuristic of the resulting line length would also do the trick - if it is too long (I'd vote for string.length>100) then split into multiple lines.

**Comment by codemanyak** [comment]
Yes. I'm preparing a method `Element.breakTextLines(int maxLineLength)` that breaks down all text lines longer than given line length parameter using the tokenized lines and preserving already placed line continuators.
This method can then be used in the root postprocessing of parsers (with a line length limit configurable in the import preferences) or via the menu for e.g. a selected element or even an entire diagram with some limit the user may enter interactively. (I just realize on dscribing it, that for latter case there should also be an opportunity to re-break the lines with a larger limit, not preserving the line continuators then.)
For a COBOL import the result might look like this (limit 80 characters):
![grafik](https://user-images.githubusercontent.com/15326471/47514752-6d3e9980-d881-11e8-9efc-db91654a50e0.png)

**Comment by GitMensch** [comment]
Allowing this both in the import and interactively on any element is a good choice!

**Comment by codemanyak** [comment]
> Allowing this both in the import and interactively on any element is a good choice!

Right. Sometimes it's really good to let an idea ripen for some time.
The import option part is done and committed to [bugfix](https://github.com/fesch/Structorizer.Desktop/commit/3a92693f9ac9b0c6a5d8a3b9806f7bb900ae2707) branch.

**Comment by codemanyak** [comment]
I think, this fix did also satisfy #446, I simply hadn't been aware then.
